syntax = "proto2";

package voting;

import "google/protobuf/timestamp.proto";

service eVotingReplica {
    rpc CreateElection (Election) returns (Status);
    rpc CastVote (Vote) returns (Status);
    rpc GetResult(ElectionName) returns (ElectionResult);
    rpc ElectionRecovery(Elections) returns (Status);
}

message Voter {
    required string name = 1;
    required string group = 2;
}
message Status {
    required int32 code = 1;
}

message Election {
    required string name = 1;
    repeated string groups = 2;
    repeated string choices = 3;
    required google.protobuf.Timestamp end_date = 4;
}
message Vote {
    required string election_name = 1;
    required string choice_name = 2;
}
message ElectionName {
    required string name = 1;
}
message VoteCount {
    required string choice_name = 1;
    required int32 count = 2;
}
message ElectionResult {
    required int32 status = 1;
    repeated VoteCount count = 2; 
}
message ElectionStatus {
    required string name = 1;
    repeated string groups = 2;
    repeated string choices = 3;
    repeated VoteCount count = 4;
    repeated string voters = 5;
    required google.protobuf.Timestamp end_date = 6;
}
message Elections {
    repeated ElectionStatus elections = 1;
}
